auth/scram.go

```diff
@@ -0,0 +1,253 @@
+	// Convert to an event batch
+	builders := evbatch.CreateColBuilders(s.userCredsSchema.EventSchema.ColumnTypes())
+	builders[0].(*evbatch.StringColBuilder).Append(username)
+	builders[1].(*evbatch.BytesColBuilder).Append(buff)
+	evBatch := evbatch.NewBatchFromBuilders(s.userCredsSchema.EventSchema, builders...)
+	processorID := s.userCredsSchema.PartitionProcessorMapping[0]
+	// Create a process batch and send it to be stored
+	processBatch := proc.NewProcessBatch(processorID, evBatch, common.UserCredsReceiverID, 0, -1)
+	return s.sendCredsBatch(processBatch)
+}
+
+func (s *ScramManager) DeleteUserCredentials(username string) error {
+	// We create a batch with just the key col
+	columnTypes := []types.ColumnType{types.ColumnTypeString}
+	schema := evbatch.NewEventSchema([]string{"username"}, columnTypes)
+	colBuilders := evbatch.CreateColBuilders(columnTypes)
+	colBuilders[0].(*evbatch.StringColBuilder).Append(username)
+	batch := evbatch.NewBatchFromBuilders(schema, colBuilders...)
+	processorID := s.userCredsSchema.PartitionProcessorMapping[0]
+	processBatch := proc.NewProcessBatch(processorID, batch, common.UserCredsDeleteReceiverID, 0, -1)
+	return s.sendCredsBatch(processBatch)
+}
+
+func CalcHash(hg scram.HashGeneratorFcn, b []byte) []byte {
+	hash := hg()
+	hash.Write(b)
+	return hash.Sum(nil)
+}
+
+func CalcHMAC(hg scram.HashGeneratorFcn, key, buff []byte) []byte {
+	hm := hmac.New(hg, key)
+	hm.Write(buff)
+	return hm.Sum(nil)
+}
+
+func (s *ScramManager) lookupCredential(username string) (scram.StoredCredentials, error) {
+	batch, err := s.executeQuerySingleResultBatch(GetCredsQueryName, []any{username})
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	user := batch.Columns[0].(*evbatch.StringColumn).Get(0)
+	if user != username {
+		// sanity check
+		return scram.StoredCredentials{}, errors.New("returned username does not match requested username")
+	}
+	credsBuff := batch.Columns[1].(*evbatch.BytesColumn).Get(0)
+	var creds userCreds
+	if err := json.Unmarshal(credsBuff, &creds); err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	var storedCreds scram.StoredCredentials
+	v, err := base64.StdEncoding.DecodeString(creds.StoredKey)
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	storedCreds.StoredKey = v
+	v, err = base64.StdEncoding.DecodeString(creds.ServerKey)
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	storedCreds.ServerKey = v
+	storedCreds.KeyFactors.Iters = creds.Iters
+	storedCreds.KeyFactors.Salt = creds.Salt
+	return storedCreds, nil
+}
+
+func (s *ScramManager) executeQuerySingleResultBatch(queryName string, args []any) (*evbatch.Batch, error) {
+	ch := make(chan *evbatch.Batch, 1)
+	_, err := s.queryManager.ExecutePreparedQueryWithHighestVersion(queryName, args, math.MaxInt64,
+		func(last bool, numLastBatches int, batch *evbatch.Batch) error {
+			if numLastBatches != 1 {
+				panic("sys query must have 1 partition")
```

Comment:
I generally only use panics for checking invariants - these are things that should never occur in production and imply something is seriously wrong. If the invariant is broken there is a serious bug and the crash is the least of our worries. I prefer to use panics than errors in these cases - otherwise we can end up with lots of errors that never actually happen and lots of unnecessary error handling code. To me, errors are for things that can happen, and don't imply a bug in the server, they are triggered by bad config, or by inputs into the system.

Reply:
> I generally only use panics for checking invariants - these are things that should never occur in production and imply something is seriously wrong. If the invariant is broken there is a serious bug and the crash is the least of our worries. I prefer to use panics than errors in these cases - otherwise we can end up with lots of errors that never actually happen and lots of unnecessary error handling code. To me, errors are for things that can happen, and don't imply a bug in the server, they are triggered by bad config, or by inputs into the system.

I totally understand your point, and I wasn't suggesting using errors. I was simply wondering if it's worth that we use a custom panic function that has a conditional recover, that way if we wanna deploy the system and disable panicking, we can do that.

It's not a very important point and I believe we're on the same page about how to handle invariants.